### ISSUE TITLE
v0.8.2 TPOTRegressor terminates prematurely

### ISSUE THREAD

**Issue by ap307** [rhiever/tpot#490]
Tried running the Boston housing price example on TPOT v0.8.2 in Python 3.6.

The fit() function runs for a few seconds, reaches a few percentage points, and then terminates. No pipeline is successfully fitted.

Code being run:

```
from tpot import TPOTRegressor
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split

digits = load_boston()
X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target,
                                                    train_size=0.75, test_size=0.25)

tpot = TPOTRegressor(generations=5, population_size=50, verbosity=2)
tpot.fit(X_train, y_train)
```

**Comment by rhiever** [comment]
Can you please provide more details on your Python environment and OS? That example is running fine on my machines.

**Comment by ap307** [comment]
Running Python 3.6.1 on Windows 10. Appreciate the help.

**Comment by ap307** [comment]
Got the Boston housing price example to work after uninstalling xgboost v0.6. Appears that it's the interaction with xgboost that is causing the issue? Had installed it using pip install xgboost.

**Comment by ap307** [comment]
OK, I think the issue is resolved. Unfortunately v0.6a2 is not available on WIndows 10 as far as I can see at this stage.

**Comment by rhiever** [comment]
TPOT has been a little finnicky with XGBoost, but the latest release should work with XGBoost 0.6a2. I just committed a patch that should add support in for 0.6. There's a weird issue where XGBoost stores both a `seed` and `random_state` parameter, both of which do the same thing, and XGBoost throws an error if they're not equal. Good times... thankfully they're fixing it soon.

BTW: You can install XGBoost on Anaconda now: `conda install py-xgboost`

**Comment by ap307** [comment]
Great, this is working now (on a source install of xgboost v0.6 with GPU support).

Any plans to enable use of XGboost's GPU processing support through TPOT? That would allow for dramatically faster hyperparameter optimization.

**Comment by rhiever** [comment]
I'm not sure what it would take to add XGBoost GPU support to TPOT. I'd imagine it entails that you can't set `n_jobs!=1`, but are there any additional contraints?

**Comment by ap307** [comment]
In XGBoost you just have to able to pass the parameters `updater='grow_gpu_hist'` or `updater='grow_gpu'`.

I'm not sure what the interaction with n_jobs is as I've only started testing out the GPU functionality, but so far am able to run on GPU while specifying  `n_jobs=-1`.

Tested it out - you can pass that parameter explicitly in TPOT (together with `n_jobs=-1`) and it doesn't appear to throw of an error and it does appear to be drawing on the GPU.

**Comment by rhiever** [comment]
Good to know. This sounds like a good topic to write up in the TPOT docs related to the [TPOT custom configuration interface](http://rhiever.github.io/tpot/using/#customizing-tpots-operators-and-parameters). Would you be willing to volunteer to draft that write-up, @ap307? :-)

**Comment by ap307** [comment]
@rhiever happy to volunteer, but let me perform some testing to make sure there aren't any negative interactions with `n_jobs` or other parameters for large data sets. For example when using `updater='grow_gpu_hist'` trees are limited to a depth of 15 so if the maximum depth is not constrained in a `tpot_config` parameter XGBoost will terminate mid-run.

**Comment by ap307** [comment]
@rhiever getting some unexpected behavior with XGBoost.

Setting XGBoost as the only estimator in `tpot_config`, and applying XGBoost to a training set of approximately 180,000 observations with ~600 features (mostly one hot encoded categorical variables) - based on the [Allstate Claims Severity](https://www.kaggle.com/c/allstate-claims-severity) Kaggle challenge - it's taking multiple hours for pipeline evaluations to start and during that time no processing is being done in the GPU.

Here are the variable definitions I'm using:

```
tpot_config = {
'xgboost.XGBRegressor':
    {
        'n_estimators': [400],
        'max_depth':[10, 12],
        'learning_rate': [0.01, 0.03],
        'subsample': [0.75, 1],
        'min_child_weight': [1,3],
        'n_jobs': [1],
        'updater': ['grow_gpu_hist']
    }
}
```

```
tpot = TPOTRegressor(generations=5, population_size=50, verbosity=3, config_dict=tpot_config,
                     scoring=evalerror, cv=2, max_eval_time_mins=15, n_jobs=1)
```

Given that XGBoost is the only allowed operator, I was expecting to see processing being shifted to the GPU almost immediately, but perhaps I'm missing something about how TPOT is configuring the pipelines.

**Comment by rhiever** [comment]
@weixuanfu2016, do you think this could be happening because we still use joblib even when `n_jobs=1`?

**Comment by weixuanfu2016** [comment]
Maybe, I need test it after coming back from vocation. 

"If you want to run XGBoost process in parallel using the fork backend for joblib/multiprocessing, you must build XGBoost without support for OpenMP by make no_omp=1"

Above is noted In this link:
https://github.com/dmlc/xgboost/blob/master/python-package/README.rst
